### PR TITLE
TYP: pandas/core/frame.py

### PR DESCRIPTION
**Pull request by arw2019**
- [ ] closes #xxxx
- [ ] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

I plan to do a pass through the whole file - putting it up in case anybody has feedback on what I have

### CODE REVIEW

File: pandas/core/frame.py [hunk 1]
```diff
@@ -1071,7 +1077,9 @@ def iterrows(self) -> Iterable[Tuple[Label, Series]]:
             s = klass(v, index=columns, name=k)
             yield k, s
 
-    def itertuples(self, index: bool = True, name: Optional[str] = "Pandas"):
+    def itertuples(
+        self, index: bool = True, name: Optional[str] = "Pandas"
+    ) -> Iterable[Tuple]:
```
Comment (jbrockmendel) [inline pandas/core/frame.py:1082]: `Iterable[NamedTuple]`?

File: pandas/core/frame.py [hunk 2]
```diff
@@ -643,7 +649,7 @@ def axes(self) -> List[Index]:
         return [self.index, self.columns]
 
     @property
-    def shape(self) -> Tuple[int, int]:
```
[the remark this answers is not in the document]
Comment (arw2019) [reply pandas/core/frame.py:646]: Fair point - yes. Will revert

**Comment by jreback** [comment]
suggest that you do reasonably sized pieces so this doesn't sit in review for too long - better multiple but smaller prs as this is a huge file

File: pandas/core/frame.py [hunk 3]
```diff
@@ -1294,7 +1301,13 @@ def __rmatmul__(self, other):
     # IO methods (to / from other formats)
 
     @classmethod
-    def from_dict(cls, data, orient="columns", dtype=None, columns=None) -> DataFrame:
+    def from_dict(
+        cls,
+        data,
+        orient: str = "columns",
+        dtype: Optional[Dtype] = None,
+        columns: Optional[List] = None,
```
Comment (jreback) [inline pandas/core/frame.py:1309]: i believe this should be Arraylike here

File: pandas/core/frame.py [hunk 4]
```diff
@@ -1440,7 +1456,7 @@ def to_numpy(
 
         return result
 
-    def to_dict(self, orient="dict", into=dict):
+    def to_dict(self, orient: str = "dict", into=dict) -> Union[Dict, List, Mapping]:
```
Comment (jreback) [inline pandas/core/frame.py:1459]: Mapping should be enough here, why did you add List?

File: pandas/core/frame.py [hunk 5]
```diff
@@ -1161,7 +1168,7 @@ def __len__(self) -> int:
         """
         return len(self.index)
 
-    def dot(self, other):
+    def dot(self, other: Union[AnyArrayLike, DataFrame]) -> FrameOrSeriesUnion:
```
Comment (jreback) [inline pandas/core/frame.py:1171]: be explict and add Series to these

File: pandas/core/frame.py [hunk 6]
```diff
@@ -1615,15 +1631,15 @@ def to_dict(self, orient="dict", into=dict):
...
-        destination_table,
-        project_id=None,
-        chunksize=None,
-        reauth=False,
-        if_exists="fail",
-        auth_local_webserver=False,
-        table_schema=None,
-        location=None,
-        progress_bar=True,
+        destination_table: str,
```
Comment (jreback) [inline pandas/core/frame.py:1634]: do these match the typings in the implementation function?

File: pandas/core/frame.py [hunk 7]
```diff
@@ -3311,7 +3329,7 @@ def _box_col_values(self, values, loc: int) -> Series:
     # ----------------------------------------------------------------------
     # Unsorted
 
-    def query(self, expr, inplace=False, **kwargs):
+    def query(self, expr: str, inplace: bool = False, **kwargs) -> Optional[DataFrame]:
```
Comment (jreback) [inline pandas/core/frame.py:3332]: i don't think this is correct for the output type

File: pandas/core/frame.py [hunk 8]
```diff
@@ -3472,10 +3490,13 @@ def query(self, expr, inplace=False, **kwargs):
...
         if inplace:
             self._update_inplace(result)
+            return None
         else:
             return result
 
-    def eval(self, expr, inplace=False, **kwargs):
+    def eval(
+        self, expr: str, inplace: bool = False, **kwargs
+    ) -> Optional[Union[AnyArrayLike, Scalar]]:
```
Comment (jreback) [inline pandas/core/frame.py:3499]: the output type here is suspect

File: pandas/core/frame.py [hunk 9]
```diff
@@ -5912,7 +5940,9 @@ def nsmallest(self, n, columns, keep="first") -> DataFrame:
             self, n=n, keep=keep, columns=columns
         ).nsmallest()
 
-    def swaplevel(self, i=-2, j=-1, axis=0) -> DataFrame:
+    def swaplevel(
+        self, i: Union[int, str] = -2, j: Union[int, str] = -1, axis: Axis = 0
```
Comment (jreback) [inline pandas/core/frame.py:5944]: these are just Axis

File: pandas/core/frame.py [hunk 10]
```diff
@@ -5943,7 +5973,9 @@ def swaplevel(self, i=-2, j=-1, axis=0) -> DataFrame:
             result.columns = result.columns.swaplevel(i, j)
         return result
 
-    def reorder_levels(self, order, axis=0) -> DataFrame:
+    def reorder_levels(
+        self, order: Union[List[int], List[str]], axis: Axis = 0
```
Comment (jreback) [inline pandas/core/frame.py:5977]: List[Axis]

File: pandas/core/frame.py [hunk 11]
```diff
@@ -7636,7 +7677,15 @@ def transform(
         assert isinstance(result, DataFrame)
         return result
 
-    def apply(self, func, axis=0, raw=False, result_type=None, args=(), **kwds):
+    def apply(
+        self,
+        func: Callable,
```
Comment (jreback) [inline pandas/core/frame.py:7682]: I believe we are more specific on this, though it maybe difficult

File: pandas/core/reshape/pivot.py [hunk 12]
```diff
@@ -612,7 +612,7 @@ def crosstab(
         margins=margins,
         margins_name=margins_name,
         dropna=dropna,
-        **kwargs,
+        **kwargs,  # type: ignore[arg-type]
```
Comment (jreback) [inline pandas/core/reshape/pivot.py:615]: don't add type ignores

**Review by jreback — CHANGES_REQUESTED** [review]
would prefer you don't mix easy and hard ones. e.g. a pass on bool / strings ones. and Axis / Level. These are easy to review. by mixing lots of things its very hard to give good scrutiny here.

would prefer that.

**Comment by jreback** [comment]
cc @simonjayhawkins @WillAyd

**Comment by arw2019** [comment]
> would prefer you don't mix easy and hard ones. e.g. a pass on bool / strings ones. and Axis / Level. These are easy to review. by mixing lots of things its very hard to give good scrutiny here.
> 
> would prefer that.

For sure. I'll keep the hard ones that I've already attempted here so all the comments are in one place. I'll resubmit easy and follow-ons in separate PRs

File: pandas/core/frame.py (hunk 8 above)
Comment (arw2019) [reply pandas/core/frame.py:3503]: Looking

File: pandas/core/frame.py (hunk 10 above)
Comment (arw2019) [reply pandas/core/frame.py:5981]: Fixed (in #38441)

File: pandas/core/frame.py (hunk 11 above)
Comment (arw2019) [reply pandas/core/frame.py:7686]: Looking - maybe we could do something like what was done for `AggFuncType`

File: pandas/core/reshape/pivot.py (hunk 12 above)
Comment (arw2019) [reply pandas/core/reshape/pivot.py:615]: okay I'll skip pivot in this PR (this is a weird issue with typing `**kwargs`)

File: pandas/core/frame.py (hunk 6 above)
Comment (arw2019) [reply pandas/core/frame.py:1638]: Yes

In fact can do more or less copy paste for the whole signature from the implementation here and in `to_html`. I might separate that out

**Comment by jreback** [comment]
needs merge master and will look

File: pandas/core/frame.py (hunk 7 above)
Comment (arw2019) [reply pandas/core/frame.py:3348]: I think this is right. From docstring:
```
DataFrame resulting from the provided query expression or None if ``inplace=True``.
```

File: pandas/core/frame.py (hunk 8 above)
Comment (arw2019) [reply pandas/core/frame.py:3515]: From docstring:
```
        Returns
        -------
        ndarray, scalar, pandas object, or None
            The result of the evaluation or None if ``inplace=True``.
```
so does `Optional[Union[AnyArrayLike, DataFrame, Scalar]]` work?

**Comment by arw2019** [comment]
mypy green + all comments addressed (or moved out to orthogonal PRs)

File: pandas/core/frame.py (hunk 4 above)
Comment (WillAyd) [reply pandas/core/frame.py:1460]: You can at least get rid of Dict then - no reason to include Dict and Mapping unless I am overlooking something

File: pandas/core/frame.py [hunk 13]
```diff
@@ -3946,15 +3970,17 @@ def _sanitize_column(self, value):
         return value
 
     @property
-    def _series(self):
+    def _series(self) -> "Dict[int, Series]":
```
Comment (WillAyd) [inline pandas/core/frame.py:3973]: Why is this quoted?

File: pandas/core/frame.py [hunk 14]
```diff
@@ -7802,7 +7828,7 @@ def apply(
         )
         return op.get_result()
 
-    def applymap(self, func, na_action: Optional[str] = None) -> DataFrame:
+    def applymap(self, func: Callable, na_action: Optional[str] = None) -> DataFrame:
```
Comment (WillAyd) [inline pandas/core/frame.py:7831]: Can this be typed more strictly than Callable? Subscripting Callable is much more useful to a reader if possible

**Review by WillAyd — CHANGES_REQUESTED** [review]
Looks pretty good - just a few things

File: pandas/core/frame.py (hunk 4 above)
Comment (arw2019) [reply pandas/core/frame.py]: Yes, that's right - done

File: pandas/core/frame.py (hunk 13 above)
Comment (arw2019) [reply pandas/core/frame.py]: It shouldn't be - quotes removed

File: pandas/core/frame.py (hunk 14 above)
Comment (arw2019) [reply pandas/core/frame.py:7831]: Yes - @jreback asked for this too. I'll put up a follow-on for this (to avoid expanding the scope of this PR)

File: pandas/core/frame.py [hunk 15]
```diff
@@ -1292,7 +1302,13 @@ def __rmatmul__(self, other):
     # IO methods (to / from other formats)
 
     @classmethod
-    def from_dict(cls, data, orient="columns", dtype=None, columns=None) -> DataFrame:
+    def from_dict(
+        cls,
+        data,
```
Comment (ivanovmg) [inline pandas/core/frame.py:1307]: Can ``data`` be typed too?

Comment (arw2019) [reply pandas/core/frame.py:1307]: For sure!

(Not here though, in the interest of refining in the scope of the PR)

Comment (jreback) [reply pandas/core/frame.py:1307]: yes these need to be smal / incremental PRs

File: pandas/core/frame.py (hunk 7 above)
Comment (jreback) [reply pandas/core/frame.py:3381]: no my point is that the doc-string is wrong. this can be a Series or Scalar. basically this can be anything. I would remove it for now.

File: pandas/core/frame.py [hunk 16]
```diff
@@ -3518,10 +3539,13 @@ def query(self, expr: str, inplace: bool = False, **kwargs):
...
         if inplace:
             self._update_inplace(result)
+            return None
         else:
             return result
 
-    def eval(self, expr: str, inplace: bool = False, **kwargs):
+    def eval(
+        self, expr: str, inplace: bool = False, **kwargs
+    ) -> Optional[Union[AnyArrayLike, DataFrame, Scalar]]:
```
Comment (jreback) [inline pandas/core/frame.py:3548]: same remove this for now leave for another pass.

File: pandas/core/frame.py [hunk 17]
```diff
@@ -6093,7 +6119,7 @@ def _arith_op(left, right):
         new_data = self._dispatch_frame_op(other, _arith_op)
         return new_data
 
-    def _construct_result(self, result) -> DataFrame:
+    def _construct_result(self, result: DataFrame) -> DataFrame:
```
Comment (jreback) [inline pandas/core/frame.py:6122]: this is not complete, this would be an ndarray, scalar or DataFrame

Comment (jreback) [reply pandas/core/frame.py:6122]: essentially anything that can be input to the constructor. leave it out for now.

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/frame.py (hunk 17 above)
Comment (arw2019) [reply pandas/core/frame.py]: Done

File: pandas/core/frame.py (hunk 16 above)
Comment (arw2019) [reply pandas/core/frame.py]: Removed

File: pandas/core/frame.py (hunk 7 above)
Comment (arw2019) [reply pandas/core/frame.py]: Ok - removed (and will fix the docstring in follow-on, too)

File: pandas/core/frame.py (hunk 1 above)
Comment (simonjayhawkins) [reply pandas/core/frame.py:1076]: I think you can put `Iterable[Tuple[Any, ...]]` so that we don't have a generic type without type parameters. Since the tuple/namedtuple is data dependent, we won't be able to refine further.

File: pandas/core/frame.py [hunk 18]
```diff
@@ -1157,7 +1163,7 @@ def __len__(self) -> int:
         """
         return len(self.index)
 
-    def dot(self, other):
+    def dot(self, other: Union[AnyArrayLike, FrameOrSeriesUnion]) -> FrameOrSeriesUnion:
```
Comment (simonjayhawkins) [inline pandas/core/frame.py:1166]: `AnyArrayLike` includes `Series`, so could probably use `Union[AnyArrayLike, DataFrame])`

will probably want to overload here in a future pass to refine the return type and avoid union return types

File: pandas/core/frame.py [hunk 19]
```diff
@@ -1290,7 +1300,13 @@ def __rmatmul__(self, other):
...
 
     @classmethod
-    def from_dict(cls, data, orient="columns", dtype=None, columns=None) -> DataFrame:
+    def from_dict(
+        cls,
+        data,
+        orient: str = "columns",
+        dtype: Optional[Dtype] = None,
+        columns: Optional[List[Label]] = None,
+    ) -> DataFrame:
```
Comment (simonjayhawkins) [inline pandas/core/frame.py:1309]: This returns same type as `cls`, so could use a typevar here in a future pass.

File: pandas/core/frame.py [hunk 20]
```diff
@@ -1369,7 +1385,10 @@ def from_dict(cls, data, orient="columns", dtype=None, columns=None) -> DataFram
         return cls(data, index=index, columns=columns, dtype=dtype)
 
     def to_numpy(
-        self, dtype=None, copy: bool = False, na_value=lib.no_default
+        self,
+        dtype: Optional[Dtype] = None,
```
Comment (simonjayhawkins) [inline pandas/core/frame.py:1389]: `dtype` is passed onto `np.array` constructor. `Dtype` includes `ExtensionDtype`, which I think would probably raise at runtime and create mypy errors when we add numpy types shortly.

maybe best to leave untyped for now pending being able to use a dtype alias from numpy shortly.

**Review by simonjayhawkins — COMMENTED** [review]
Thanks @arw2019 for the PR. some initial comments, i've not fully reviewed this PR yet.

File: pandas/core/frame.py [hunk 21]
```diff
@@ -2708,7 +2727,7 @@ def info(
             show_counts=show_counts,
         )
 
-    def memory_usage(self, index=True, deep=False) -> Series:
+    def memory_usage(self, index: bool = True, deep: bool = False) -> Series:
```
Comment (simonjayhawkins) [inline pandas/core/frame.py:2730]: return type is `self._constructor_sliced`. The only way to tie the result type to a class attribute is to make DataFrame/Series generic. out of scope for now. see #32638

File: pandas/core/frame.py [hunk 22]
```diff
@@ -2931,7 +2950,7 @@ def T(self) -> DataFrame:
     # ----------------------------------------------------------------------
     # Indexing Methods
 
-    def _ixs(self, i: int, axis: int = 0):
+    def _ixs(self, i: int, axis: Axis = 0):
```
Comment (simonjayhawkins) [inline pandas/core/frame.py:2953]: `Axis` includes `str`, we can only use the `Axis` alias if we use `axis = self._get_axis_number(axis)` or pass it onto a method that does so. Here we don't do that lookup so should use `int`

File: pandas/core/frame.py [hunk 23]
```diff
@@ -3287,7 +3306,9 @@ def _set_item(self, key, value):
 
         self._set_item_mgr(key, value)
 
-    def _set_value(self, index, col, value, takeable: bool = False):
+    def _set_value(
+        self, index: int, col, value: Scalar, takeable: bool = False
```
Comment (simonjayhawkins) [inline pandas/core/frame.py:3310]: the docstring states `index : row label`, is `index: int` correct here?

maybe need to update the docstring to the
```
parameter: type
    description
```
format

File: pandas/core/frame.py [hunk 24]
```diff
@@ -3944,15 +3966,17 @@ def _sanitize_column(self, value):
         return value
 
     @property
-    def _series(self):
+    def _series(self) -> Dict[int, Series]:
```
Comment (simonjayhawkins) [inline pandas/core/frame.py:3969]: `item` in `idx, item in enumerate(self.columns)` is not necessarily an `int`

File: pandas/core/frame.py [hunk 25]
```diff
@@ -6094,7 +6118,7 @@ def _arith_op(left, right):
         new_data = self._dispatch_frame_op(other, _arith_op)
         return new_data
 
-    def _construct_result(self, result) -> DataFrame:
```
Comment (simonjayhawkins) [inline pandas/core/frame.py:6097]: why was this deleted? self._constructor is typed as `Type[DataFrame]`. while not 100% correct, this should not be giving mypy errors.

File: pandas/core/frame.py [hunk 26]
```diff
@@ -8268,7 +8292,9 @@ def merge(
             validate=validate,
         )
 
-    def round(self, decimals=0, *args, **kwargs) -> DataFrame:
+    def round(
+        self, decimals: Union[int, Dict, Series] = 0, *args, **kwargs
```
Comment (simonjayhawkins) [inline pandas/core/frame.py:8296]: try to avoid generics without type parameters. `Dict[Label, int]`?

File: pandas/core/frame.py [hunk 27]
```diff
@@ -8382,7 +8408,9 @@ def _series_round(s, decimals):
     # ----------------------------------------------------------------------
     # Statistical methods, etc.
 
-    def corr(self, method="pearson", min_periods=1) -> DataFrame:
+    def corr(
+        self, method: Union[str, Callable] = "pearson", min_periods: int = 1
```
Comment (simonjayhawkins) [inline pandas/core/frame.py:8412]: Callable[[np.ndarray,np.ndarray], float]?

File: pandas/core/frame.py [hunk 28]
```diff
@@ -9459,7 +9493,7 @@ def to_period(self, freq=None, axis: Axis = 0, copy: bool = True) -> DataFrame:
         setattr(new_obj, axis_name, new_ax)
         return new_obj
 
-    def isin(self, values) -> DataFrame:
+    def isin(self, values: Union[FrameOrSeriesUnion, Dict, Iterable]) -> DataFrame:
```
Comment (simonjayhawkins) [inline pandas/core/frame.py:9496]: type parameters for Dict and Iterable?

File: pandas/core/frame.py (hunk 18 above)
Comment (arw2019) [reply pandas/core/frame.py:1166]: opened #38765 to add the overload, will revert here to orthogonalize the PRs

File: pandas/core/frame.py (hunk 20 above)
Comment (arw2019) [reply pandas/core/frame.py]: I see we recently added `NpDtype` to `pd._typing`. Shall I use that here?

File: pandas/core/frame.py [hunk 29]
```diff
@@ -1436,7 +1455,7 @@ def to_numpy(
 
         return result
 
-    def to_dict(self, orient: str = "dict", into=dict):
+    def to_dict(self, orient: str = "dict", into=dict) -> Union[List, Mapping]:
```
[the remark this answers is not in the document]
Comment (arw2019) [reply pandas/core/frame.py:1438]: Will open a dedicated PR

File: pandas/core/frame.py [hunk 30]
```diff
@@ -1725,7 +1744,7 @@ def from_records(
         exclude=None,
         columns=None,
         coerce_float: bool = False,
-        nrows=None,
+        nrows: Optional[bool] = None,
```
[the remark this answers is not in the document]
Comment (arw2019) [reply pandas/core/frame.py]: yes (made a mistake)

File: pandas/core/frame.py (hunk 21 above)
Comment (arw2019) [reply pandas/core/frame.py]: Ok - reverted

File: pandas/core/frame.py (hunk 22 above)
Comment (arw2019) [reply pandas/core/frame.py]: Right, fixed

File: pandas/core/frame.py (hunk 23 above)
Comment (arw2019) [reply pandas/core/frame.py]: right, so should be `Label` I think (again, my bad)

Docstring updated.

File: pandas/core/frame.py (hunk 24 above)
Comment (arw2019) [reply pandas/core/frame.py]: Right, should be `Label` I think

File: pandas/core/frame.py (hunk 25 above)
Comment (arw2019) [reply pandas/core/frame.py]: I think by mistake, reverting

File: pandas/core/frame.py (hunk 26 above)
Comment (arw2019) [reply pandas/core/frame.py]: yes - added

File: pandas/core/frame.py (hunk 28 above)
Comment (arw2019) [reply pandas/core/frame.py]: Added

**Review by arw2019 — COMMENTED** [review]
Thanks @simonjayhawkins for the detailed review! I addressed the comments (all of them, I believe) and will submit follow-on PRs for where you suggested doing a second pass

File: pandas/core/frame.py (hunk 29 above)
[the remark this answers is not in the document]
Comment (jreback) [reply pandas/core/frame.py:1438]: hmm this returns a dict right? how is it a LIst union?

File: pandas/core/frame.py [hunk 31]
```diff
@@ -9395,7 +9424,11 @@ def resample(
         )
 
     def to_timestamp(
-        self, freq=None, how: str = "start", axis: Axis = 0, copy: bool = True
+        self,
+        freq: Optional[str] = None,
```
Comment (jreback) [inline pandas/core/frame.py:9428]: this is a frequency type, I think we have a type for it

File: pandas/core/frame.py [hunk 32]
```diff
@@ -9428,7 +9461,9 @@ def to_timestamp(
         setattr(new_obj, axis_name, new_ax)
         return new_obj
 
-    def to_period(self, freq=None, axis: Axis = 0, copy: bool = True) -> DataFrame:
+    def to_period(
+        self, freq: Optional[str] = None, axis: Axis = 0, copy: bool = True
```
Comment (jreback) [inline pandas/core/frame.py:9465]: same here

File: pandas/core/frame.py [hunk 33]
```diff
@@ -9460,7 +9495,12 @@ def to_period(self, freq=None, axis: Axis = 0, copy: bool = True) -> DataFrame:
         setattr(new_obj, axis_name, new_ax)
         return new_obj
 
-    def isin(self, values) -> DataFrame:
+    def isin(
+        self,
+        values: Union[
```
Comment (jreback) [inline pandas/core/frame.py:9500]: leave this off for now. i am not sure this is right and this is complicated.

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/frame.py (hunk 32 above)
Comment (arw2019) [reply pandas/core/frame.py]: added

File: pandas/core/frame.py (hunk 31 above)
Comment (arw2019) [reply pandas/core/frame.py]: I don't think we do in typing, but I made one

File: pandas/core/frame.py (hunk 29 above)
[the remark this answers is not in the document]
Comment (arw2019) [reply pandas/core/frame.py]: It depends on orient value. I'm working on a PR with the `Literal` workaround so will revert here for now

File: pandas/core/frame.py (hunk 33 above)
Comment (arw2019) [reply pandas/core/frame.py]: Done

**Comment by arw2019** [comment]
Green + addressed comments

File: pandas/core/frame.py (hunk 20 above)
Comment (simonjayhawkins) [reply pandas/core/frame.py]: I have reservations on using NpDtype (and adding types to dtype parameters in general, such as #38808) as while these types resolve to `Any`, mypy will not report any inconsistencies.

The issue here will be picked up when we have the numpy types and therefore the burden on reviewers is reduced.

My reservations on adding these ahead of time is that there will likely be more errors to resolve when we transition to using numpy types. Although that boat has already sailed, xref #36092

so rather than doing anything more here now, I'm happy to accept that mypy is green.

**Review by simonjayhawkins — APPROVED** [review]
Thanks @arw2019 lgtm

File: pandas/core/frame.py (hunk 20 above)
Comment (arw2019) [reply pandas/core/frame.py]: Ok. Removed `NpDtype` usage here (much easier to revisit after #36092, where there seems to be a lot to do already, has gone in)

**Review by jreback — APPROVED** [review]

File: pandas/_typing.py [hunk 34]
```diff
@@ -84,6 +84,7 @@
 FrameOrSeries = TypeVar("FrameOrSeries", bound="NDFrame")
 
 Axis = Union[str, int]
+Frequency = Optional[str]
```
Comment (jreback) [inline pandas/_typing.py:87]: this also includes DateOffset

File: pandas/core/frame.py [hunk 35]
```diff
@@ -4558,7 +4580,11 @@ def _replace_columnwise(
 
     @doc(NDFrame.shift, klass=_shared_doc_kwargs["klass"])
     def shift(
-        self, periods=1, freq=None, axis: Axis = 0, fill_value=lib.no_default
+        self,
+        periods=1,
+        freq: Frequency = None,
```
Comment (jreback) [inline pandas/core/frame.py:4585]: should be Optional

File: pandas/core/frame.py [hunk 36]
```diff
@@ -9395,7 +9427,11 @@ def resample(
         )
 
     def to_timestamp(
-        self, freq=None, how: str = "start", axis: Axis = 0, copy: bool = True
+        self,
+        freq: Frequency = None,
```
Comment (jreback) [inline pandas/core/frame.py:9431]: optional

File: pandas/core/frame.py [hunk 37]
```diff
@@ -9428,7 +9464,9 @@ def to_timestamp(
         setattr(new_obj, axis_name, new_ax)
         return new_obj
 
-    def to_period(self, freq=None, axis: Axis = 0, copy: bool = True) -> DataFrame:
+    def to_period(
+        self, freq: Frequency = None, axis: Axis = 0, copy: bool = True
```
Comment (jreback) [inline pandas/core/frame.py:9468]: same

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by jreback** [comment]
@arw2019 if you want to merge master and update (or close and open a new one ok too)

File: pandas/_typing.py (hunk 34 above)
Comment (simonjayhawkins) [reply pandas/_typing.py:87]: maybe better to also remove the optional here, xref https://github.com/pandas-dev/pandas/pull/38416/files#r552673889

**Comment by arw2019** [comment]
@jreback @simonjayhawkins updated, CI green, comments addressed

File: pandas/core/frame.py [hunk 38]
```diff
@@ -3373,16 +3381,21 @@ def _set_item(self, key, value):
 
         self._set_item_mgr(key, value)
 
-    def _set_value(self, index, col, value, takeable: bool = False):
+    def _set_value(
+        self, index: IndexLabel, col, value: Scalar, takeable: bool = False
```
Comment (simonjayhawkins) [inline pandas/core/frame.py:3385]: signature and docstring params inconsistent

File: pandas/core/frame.py [hunk 39]
```diff
@@ -3373,16 +3381,21 @@ def _set_item(self, key, value):
...
+    ) -> None:
         """
         Put single value at passed column and index.
 
         Parameters
         ----------
-        index : row label
-        col : column label
+        index : Label
```
Comment (simonjayhawkins) [inline pandas/core/frame.py:3392]: even though this function is internal, the docstring should not use typing alias, but numpydoc conventions

File: pandas/core/frame.py [hunk 40]
```diff
@@ -3373,16 +3381,21 @@ def _set_item(self, key, value):
...
+            row label
+        col : Label
+            column label
         value : scalar
-        takeable : interpret the index/col as indexers, default False
+        takeable : bool, default False
+            Sets whether or not index/col interpreted as indexers
```
Comment (simonjayhawkins) [inline pandas/core/frame.py:3398]: I thought we had a check that the descriptions end in a period.

File: pandas/core/frame.py [hunk 41]
```diff
@@ -4049,15 +4063,17 @@ def _sanitize_column(self, value):
         return value
 
     @property
-    def _series(self):
+    def _series(self) -> Dict[IndexLabel, Series]:
```
Comment (simonjayhawkins) [inline pandas/core/frame.py:4066]: IndexLabel is Union[Hashable, Sequence[Hashable]]

the dictionary keys will be the items from Index.__iter__

so a tuple for a MultiIndex, which is a Sequence but we should be more precise for return types

is the return type of Index.__iter__ typed? If so should be the same. or sometimes better to type lower level functions first.

this module is difficult to type since it is high level and user facing, probably should be one of the last modules to type

File: pandas/core/frame.py [hunk 42]
```diff
@@ -2786,7 +2794,7 @@ def info(
             show_counts=show_counts,
         )
 
-    def memory_usage(self, index=True, deep=False) -> Series:
```
Comment (simonjayhawkins) [inline pandas/core/frame.py:2789]: why is the return type removed?

File: pandas/core/frame.py [hunk 43]
```diff
@@ -9598,7 +9620,10 @@ def to_period(
         setattr(new_obj, axis_name, new_ax)
         return new_obj
 
-    def isin(self, values) -> DataFrame:
+    def isin(
+        self,
+        values,
+    ) -> DataFrame:
```
Comment (simonjayhawkins) [inline pandas/core/frame.py:9626]: can lose the rogue comma

File: pandas/core/frame.py (hunk 42 above)
Comment (arw2019) [reply pandas/core/frame.py:2789]: it shouldn't be - reverted

File: pandas/core/frame.py (hunk 41 above)
Comment (arw2019) [reply pandas/core/frame.py]: okay will remove for now

File: pandas/core/frame.py (hunk 43 above)
Comment (arw2019) [reply pandas/core/frame.py]: done

File: pandas/core/frame.py (hunk 42 above)
Comment (arw2019) [reply pandas/core/frame.py:2789]: actually - it's from your previous comment 
https://github.com/pandas-dev/pandas/pull/38416/files#r549454789

File: pandas/core/frame.py (hunk 4 above)
Comment (simonjayhawkins) [reply pandas/core/frame.py]: > with `orient="records"` we return a list:

we should probably overload using Literal to avoid Union return types. but not in this pass.

**Comment by simonjayhawkins** [comment]
> @jreback @simonjayhawkins updated, CI green, comments addressed

ok. maybe we should merge this. there has been plenty of discussion here.

**Review by simonjayhawkins — APPROVED** [review]

**Comment by simonjayhawkins** [comment]
@jreback @WillAyd i'm merging later today if no objections.

mypy is green and typing PRs easily become stale or get abandoned.

There are no ignores, casts or asserts added here and no existing annotations changed/deleted.

if any of the typing additions added here are problematic they will be picked up by mypy as more types are added.

**Comment by jreback** [comment]
sure let's make sure this is rebased first

**Comment by simonjayhawkins** [comment]
yep did that locally and ran mypy before I commented. will push and merge on green

**Review by jreback — APPROVED** [review]
lgtm

**Comment by simonjayhawkins** [comment]
Thanks @arw2019
